File: clase 2/PYTHON-OPCION/2_consolidar_wos_scopus.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# ============================================================================
# Fusiona archivos WoS y Scopus (ya convertido a formato WoS) en un unico
# archivo, detectando y eliminando registros duplicados.
#
# Correcciones aplicadas para compatibilidad con Tree of Science:
#   1. Cabecera FN = "Clarivate Analytics Web of Science"
#   2. Linea en blanco despues de cada ER (separador de registros)
#   3. Referencias CR de Scopus reformateadas al estilo WoS:
#      WoS:    Autor, Year, JOURNAL_ABBREV, Vvol, Ppage, DOI xxx
#      Scopus: Autor, Year, TITULO..., JOURNAL..., Vvol, Ppage
#      -> Quita el titulo, conserva solo Autor, Year, Journal, Vvol, Ppage
#
# Uso:
#   python 2_consolidar_wos_scopus.py
#
# Salidas:
#   - wos_scopus_consolidado.txt -> Archivo unificado sin duplicados
#   - reporte_consolidado.txt    -> Reporte resumen con detalle de duplicados
# ============================================================================

import os
import re
import unicodedata
from difflib import SequenceMatcher

# ============================================================
# CONFIGURACION - Editar rutas segun necesidad
# ============================================================
WOS_FILE      = "input/wos.txt"
SCOPUS_FILE   = "output/scopus_formato_wos.txt"
OUTPUT_FOLDER = "output"
OUTPUT_FILE   = os.path.join(OUTPUT_FOLDER, "wos_scopus_consolidado.txt")
REPORT_FILE   = os.path.join(OUTPUT_FOLDER, "reporte_consolidado.txt")

TITLE_SIMILARITY_THRESHOLD = 0.85
PRIORITY = "wos"   # "wos" o "scopus" - cual conservar en caso de duplicado
# ============================================================


def string_similarity(a, b):
    """Similitud de cadenas equivalente a difflib.SequenceMatcher.ratio."""
    if a == "" and b == "":
        return 1.0
    if a == "" or b == "":
        return 0.0
    return SequenceMatcher(None, a, b).ratio()


def parse_wos_records(filepath):
    """Parsea un archivo en formato WoS etiquetado y devuelve lista de dicts."""
    with open(filepath, "r", encoding="utf-8-sig") as f:
        lines = f.read().splitlines()

    records = []
    current = {}
    current_tag = None

    for raw_line in lines:
        line = raw_line.rstrip("\r")

        if line.startswith("FN ") or line.startswith("VR ") or line.startswith("EF"):
            continue

        # Verificar si es una etiqueta WoS: 2 mayusculas seguidas de un espacio
        is_tag = (len(line) >= 3
                  and line[2] == " "
                  and re.match(r"^[A-Z][A-Z0-9]$", line[:2]))

        if is_tag:
            tag = line[:2]
            value = line[3:]

            if tag == "PT":
                if current:
                    records.append(current)
                current = {"PT": value}
                current_tag = "PT"
            elif tag == "ER":
                if current:
                    records.append(current)
                current = {}
                current_tag = None
            else:
                if tag in current:
                    current[tag] = current[tag] + "\n" + value
                else:
                    current[tag] = value
                current_tag = tag

        elif line.startswith("   ") and current_tag is not None and current:
            # Linea de continuacion
            current[current_tag] = current[current_tag] + "\n" + line[3:]

    if current:
        records.append(current)

    return records


def normalize_doi(doi):
    """Normaliza DOI para comparacion."""
    if not doi:
        return None
    doi = doi.strip().lower()
    doi = re.sub(r"^https?://(dx\.)?doi\.org/", "", doi)
    return doi


def normalize_title(title):
    """Normaliza titulo: minusculas, sin acentos, sin puntuacion, espacios colapsados."""
    if not title:
        return ""
    t = re.sub(r"\s+", " ", title).lower()
    # Quitar acentos
    t = unicodedata.normalize("NFKD", t)
    t = "".join(c for c in t if not unicodedata.combining(c))
    # Quitar caracteres no alfanumericos ni espacios
    t = re.sub(r"[^\w\s]", " ", t, flags=re.UNICODE)
    t = re.sub(r"\s+", " ", t).strip()
    return t


def reformat_scopus_cr(cr_text):
    """Reformatea las referencias CR provenientes de Scopus al estilo WoS."""
    if not cr_text:
        return cr_text

    lines = cr_text.split("\n")
    reformatted = []

    for line in lines:
        line = line.strip()
        if line == "":
            continue

        # Buscar el patron: ..., Vdigits, ...
        v_match = re.search(r",\s*(V\d+)", line)

        if v_match:
            before_v = line[:v_match.start()]
            from_v = line[v_match.start():]

            parts = [p.strip() for p in before_v.split(",")]

            # Buscar anio (numero de 4 digitos)
            year_idx = None
            year_val = None
            for idx, part in enumerate(parts):
                if re.match(r"^\d{4}$", part.strip()):
                    year_val = part.strip()
                    year_idx = idx
                    break

            if year_idx is not None:
                # Autor = todo antes del anio
                author = ", ".join(parts[:year_idx]).strip()

                # Journal = ultimo segmento despues del anio (justo antes de Vxx)
                after_year = parts[year_idx + 1:] if year_idx < len(parts) - 1 else []
                journal = after_year[-1].strip() if after_year else None

                # Limpiar from_v
                from_v_clean = re.sub(r"^,\s*", "", from_v).strip()

                if author and journal:
                    new_cr = f"{author}, {year_val}, {journal}, {from_v_clean}"
                elif author:
                    new_cr = f"{author}, {year_val}, {from_v_clean}"
                else:
                    new_cr = line
                reformatted.append(new_cr)
            else:
                reformatted.append(line)
        else:
            reformatted.append(line)

    return "\n".join(reformatted)


def is_scopus_record(rec):
    """Indica si el registro proviene de Scopus segun el campo UT."""
    ut = rec.get("UT")
    if not ut:
        return False
    return "SCOPUS:" in ut


def record_to_wos_text(rec):
    """Convierte un dict de registro a texto en formato WoS etiquetado."""
    lines = []
    lines.append("PT " + rec.get("PT", "J"))

    field_order = [
        "AU", "AF", "TI", "SO", "LA", "DT", "DE", "ID", "AB",
        "C1", "C3", "RP", "EM", "RI", "OI",
        "FU", "FX", "CR", "NR", "TC", "Z9", "U1", "U2",
        "PU", "PI", "PA", "SN", "EI", "BN",
        "J9", "JI", "PD", "PY", "VL", "IS", "SI",
        "BP", "EP", "AR", "DI", "EA", "PG",
        "PM", "WC", "WE", "SC", "GA", "UT", "OA", "HC", "HP", "DA"
    ]

    written = {"PT"}

    for tag in field_order:
        if tag in rec:
            value = rec[tag]

            # Reformatear CR para registros de origen Scopus
            if tag == "CR" and is_scopus_record(rec):
                value = reformat_scopus_cr(value)

            value_lines = value.split("\n")
            lines.append(f"{tag} {value_lines[0]}")
            for vl in value_lines[1:]:
                lines.append(f"   {vl}")
            written.add(tag)

    # Escribir cualquier etiqueta restante fuera del orden predefinido
    for tag, value in rec.items():
        if tag not in written:
            value_lines = value.split("\n")
            lines.append(f"{tag} {value_lines[0]}")
            for vl in value_lines[1:]:
                lines.append(f"   {vl}")

    lines.append("ER")
    return "\n".join(lines)


def merge_records(wos_records, scopus_records, threshold=0.85, priority="wos"):
    """Fusiona registros detectando duplicados por DOI y similitud de titulo."""
    # Construir indice por DOI para registros WoS
    wos_by_doi = {}
    for rec in wos_records:
        doi = normalize_doi(rec.get("DI"))
        if doi:
            wos_by_doi[doi] = rec

    # Lista de titulos WoS
    wos_titles = [(normalize_title(rec.get("TI", "")), rec) for rec in wos_records]

    merged = list(wos_records)
    duplicates = []
    added_from_scopus = 0
    skipped_doi = 0
    skipped_title = 0

    for s_rec in scopus_records:
        s_doi = normalize_doi(s_rec.get("DI"))
        is_dup = False

        # Verificar coincidencia por DOI
        if s_doi and s_doi in wos_by_doi:
            w_rec = wos_by_doi[s_doi]
            duplicates.append({
                "wos": w_rec,
                "scopus": s_rec,
                "match_type": "DOI",
                "score": 1.0
            })
            skipped_doi += 1
            is_dup = True

        # Verificar similitud por titulo
        if not is_dup:
            s_title = normalize_title(s_rec.get("TI", ""))
            if s_title:
                best_score = 0
                best_match = None

                for w_title, w_rec in wos_titles:
                    if not w_title:
                        continue
                    len_ratio = len(s_title) / max(len(w_title), 1)
                    if len_ratio < 0.7 or len_ratio > 1.4:
                        continue
                    score = string_similarity(s_title, w_title)
                    if score > best_score:
                        best_score = score
                        best_match = w_rec

                if best_score >= threshold:
                    duplicates.append({
                        "wos": best_match,
                        "scopus": s_rec,
                        "match_type": "TITLE",
                        "score": best_score
                    })
                    skipped_title += 1
                    is_dup = True

        if not is_dup:
            merged.append(s_rec)
            added_from_scopus += 1

    stats = {
        "wos_total":         len(wos_records),
        "scopus_total":      len(scopus_records),
        "duplicates_doi":    skipped_doi,
        "duplicates_title":  skipped_title,
        "added_from_scopus": added_from_scopus,
        "merged_total":      len(merged),
        "priority":          priority
    }

    return merged, duplicates, stats


def generate_report(stats, duplicates, report_path):
    """Genera el reporte de texto con el detalle de los duplicados."""
    lines = []

    sep_eq = "=" * 70
    sep_dash = "-" * 70

    lines.append(sep_eq)
    lines.append("  REPORTE DE CONSOLIDACION Y DESDUPLICACION")
    lines.append(sep_eq)
    lines.append("")
    lines.append(f"  Registros WoS:                  {stats['wos_total']}")
    lines.append(f"  Registros Scopus:               {stats['scopus_total']}")
    lines.append(f"  Total antes de desduplicar:     {stats['wos_total'] + stats['scopus_total']}")
    lines.append("")
    lines.append(f"  Duplicados por DOI:             {stats['duplicates_doi']}")
    lines.append(f"  Duplicados por similitud titulo:{stats['duplicates_title']}")
    lines.append(f"  Total duplicados eliminados:    {stats['duplicates_doi'] + stats['duplicates_title']}")
    lines.append("")
    lines.append(f"  Unicos anadidos desde Scopus:   {stats['added_from_scopus']}")
    lines.append(f"  Prioridad en duplicado:         {stats['priority'].upper()}")
    lines.append("")
    lines.append(f"  REGISTROS FINALES:              {stats['merged_total']}")
    lines.append(sep_eq)
    lines.append("")

    if duplicates:
        lines.append(sep_dash)
        lines.append("  DETALLE DE DUPLICADOS")
        lines.append(sep_dash)
        lines.append("")

        for i, dup in enumerate(duplicates, 1):
            w_title = re.sub(r"\s+", " ", dup["wos"].get("TI", "N/A"))[:100]
            s_title = re.sub(r"\s+", " ", dup["scopus"].get("TI", "N/A"))[:100]
            doi = dup["wos"].get("DI") or dup["scopus"].get("DI") or "N/A"

            match_info = f"  [{i:3d}] Match: {dup['match_type']}"
            if dup["match_type"] == "TITLE":
                match_info += f" (similitud: {dup['score'] * 100:.2f}%)"
            lines.append(match_info)
            lines.append(f"        DOI:    {doi}")
            lines.append(f"        WoS:    {w_title}")
            lines.append(f"        Scopus: {s_title}")
            lines.append("")

    # Escribir con BOM
    with open(report_path, "wb") as f:
        f.write(b"\xEF\xBB\xBF")
        f.write("\n".join(lines).encode("utf-8"))

    return report_path


# ============================================================
# MAIN
# ============================================================
def main():
    os.makedirs(OUTPUT_FOLDER, exist_ok=True)

    print(f"Cargando archivo WoS: {WOS_FILE}")
    wos_records = parse_wos_records(WOS_FILE)
    print(f"  -> {len(wos_records)} registros\n")

    print(f"Cargando archivo Scopus: {SCOPUS_FILE}")
    scopus_records = parse_wos_records(SCOPUS_FILE)
    print(f"  -> {len(scopus_records)} registros\n")

    print("Fusionando y desduplicando...")
    merged, duplicates, stats = merge_records(
        wos_records, scopus_records,
        threshold=TITLE_SIMILARITY_THRESHOLD,
        priority=PRIORITY
    )

    # Escribir archivo fusionado con formato WoS exacto
    print(f"\nEscribiendo archivo consolidado: {OUTPUT_FILE}")
    with open(OUTPUT_FILE, "wb") as f:
        f.write(b"\xEF\xBB\xBF")  # BOM
        f.write("FN Clarivate Analytics Web of Science\n".encode("utf-8"))
        f.write("VR 1.0\n".encode("utf-8"))

        for rec in merged:
            wos_text = record_to_wos_text(rec)
            # Linea en blanco despues de cada ER (compatibilidad Tree of Science)
            f.write((wos_text + "\n\n").encode("utf-8"))

        f.write("EF\n".encode("utf-8"))

    report_path = generate_report(stats, duplicates, REPORT_FILE)

    # Validacion
    with open(OUTPUT_FILE, "r", encoding="utf-8-sig") as f:
        content = f.read()
    pt_count = len(re.findall(r"\nPT ", content))
    er_count = len(re.findall(r"\nER\n", content))

    print()
    sep = "=" * 50
    print(sep)
    print("  RESUMEN DE CONSOLIDACION")
    print(sep)
    print(f"  Registros WoS:            {stats['wos_total']}")
    print(f"  Registros Scopus:         {stats['scopus_total']}")
    print(f"  Duplicados (DOI):         {stats['duplicates_doi']}")
    print(f"  Duplicados (titulo):      {stats['duplicates_title']}")
    print(f"  Unicos desde Scopus:      {stats['added_from_scopus']}")
    print(f"  TOTAL CONSOLIDADO:        {stats['merged_total']}")
    print(f"  Validacion: PT={pt_count}, ER={er_count}")
    print(sep)
    print(f"\n  Salida:  {OUTPUT_FILE}")
    print(f"  Reporte: {report_path}")


if __name__ == "__main__":
    main()

File: clase 2/RSTUDIO-OPCION/2_consolidar_wos_scopus.R
#!/usr/bin/env Rscript
# ============================================================================
# Merge WoS and Scopus (converted to WoS format) files into a single file,
# detecting and removing duplicate records.
#
# Fixes applied for Tree of Science compatibility:
#   1. FN header = "Clarivate Analytics Web of Science"
#   2. Blank line after each ER (record separator)
#   3. Scopus CR references reformatted to WoS style:
#      WoS:    Author, Year, JOURNAL_ABBREV, Vvol, Ppage, DOI xxx
#      Scopus: Author, Year, TITLE..., JOURNAL..., Vvol, Ppage
#      -> Removes title, keeps only Author, Year, Journal, Vvol, Ppage
#
# Usage:
#   Rscript merge_wos_scopus.R
#
# Outputs:
#   - merged_wos_scopus.txt   -> Unified file without duplicates
#   - merge_report.txt        -> Summary report with duplicate details
# ============================================================================

# ============================================================
# CONFIGURATION - Edit these paths as needed
# ============================================================
WOS_FILE      <- "input/wos.txt"
SCOPUS_FILE   <- "output/scopus_formato_wos.txt"
OUTPUT_FOLDER <- "output"
OUTPUT_FILE   <- file.path(OUTPUT_FOLDER, "wos_scopus_consolidado.txt")
REPORT_FILE   <- file.path(OUTPUT_FOLDER, "reporte_consolidado.txt")

TITLE_SIMILARITY_THRESHOLD <- 0.85
PRIORITY <- "wos"   # "wos" or "scopus" - which to keep on duplicate
# ============================================================


# ---- String similarity (equivalent to SequenceMatcher.ratio) ----
# Uses the longest common subsequence approach similar to Python's difflib
string_similarity <- function(a, b) {
  if (a == "" && b == "") return(1.0)
  if (a == "" || b == "") return(0.0)
  
  # Use adist (Levenshtein) to approximate SequenceMatcher ratio
  
  # ratio = 1 - (edit_distance / max_length)
  # This is a close approximation to SequenceMatcher for dedup purposes
  d <- adist(a, b)[1, 1]
  max_len <- max(nchar(a), nchar(b))
  return(1 - d / max_len)
}


parse_wos_records <- function(filepath) {
  lines <- readLines(filepath, encoding = "UTF-8", warn = FALSE)
  # Remove BOM if present
  if (length(lines) > 0) {
    lines[1] <- sub("^\uFEFF", "", lines[1])
  }
  
  records <- list()
  current <- list()
  current_tag <- NULL
  
  for (line in lines) {
    line <- sub("\\r$", "", line)  # remove trailing CR if any
    
    if (startsWith(line, "FN ") || startsWith(line, "VR ") || startsWith(line, "EF")) {
      next
    }
    
    # Check for a WoS tag: two uppercase chars followed by a space at position 3
    if (nchar(line) >= 3 && substr(line, 3, 3) == " " &&
        grepl("^[A-Z][A-Z0-9]$", substr(line, 1, 2))) {
      
      tag   <- substr(line, 1, 2)
      value <- substr(line, 4, nchar(line))
      
      if (tag == "PT") {
        if (length(current) > 0) {
          records <- c(records, list(current))
        }
        current <- list(PT = value)
        current_tag <- "PT"
        
      } else if (tag == "ER") {
        if (length(current) > 0) {
          records <- c(records, list(current))
        }
        current <- list()
        current_tag <- NULL
        
      } else {
        if (!is.null(current[[tag]])) {
          current[[tag]] <- paste0(current[[tag]], "\n", value)
        } else {
          current[[tag]] <- value
        }
        current_tag <- tag
      }
      
    } else if (startsWith(line, "   ") && !is.null(current_tag) && length(current) > 0) {
      # Continuation line
      current[[current_tag]] <- paste0(current[[current_tag]], "\n", substr(line, 4, nchar(line)))
    }
  }
  
  if (length(current) > 0) {
    records <- c(records, list(current))
  }
  
  return(records)
}


normalize_doi <- function(doi) {
  if (is.null(doi) || doi == "") return(NA_character_)
  doi <- trimws(tolower(doi))
  doi <- sub("^https?://(dx\\.)?doi\\.org/", "", doi)
  return(doi)
}


normalize_title <- function(title) {
  if (is.null(title) || title == "") return("")
  t <- gsub("\\s+", " ", title)
  t <- tolower(t)
  # Remove accents
  t <- iconv(t, to = "ASCII//TRANSLIT")
  if (is.na(t)) t <- tolower(gsub("\\s+", " ", title))
  # Remove non-word/non-space chars
  t <- gsub("[^\\w\\s]", " ", t, perl = TRUE)
  t <- gsub("\\s+", " ", trimws(t))
  return(t)
}


reformat_scopus_cr <- function(cr_text) {
  if (is.null(cr_text) || cr_text == "") return(cr_text)
  
  lines <- strsplit(cr_text, "\n")[[1]]
  reformatted <- character(0)
  
  for (line in lines) {
    line <- trimws(line)
    if (line == "") next
    
    # Try to find the pattern: ..., Vdigits, ...
    v_match <- regexpr(",\\s*(V\\d+)", line, perl = TRUE)
    
    if (v_match > 0) {
      before_v <- substr(line, 1, v_match - 1)
      from_v   <- substr(line, v_match, nchar(line))
      
      parts <- trimws(strsplit(before_v, ",")[[1]])
      
      # Find year (4-digit number)
      year_idx <- NULL
      year_val <- NULL
      for (idx in seq_along(parts)) {
        if (grepl("^\\d{4}$", trimws(parts[idx]))) {
          year_val <- trimws(parts[idx])
          year_idx <- idx
          break
        }
      }
      
      if (!is.null(year_idx)) {
        # Author = everything before year
        author <- paste(parts[1:(year_idx - 1)], collapse = ", ")
        author <- trimws(author)
        
        # Journal = last segment after year (just before Vxx)
        after_year <- if (year_idx < length(parts)) parts[(year_idx + 1):length(parts)] else character(0)
        journal <- if (length(after_year) > 0) trimws(after_year[length(after_year)]) else NULL
        
        # Clean from_v
        from_v_clean <- trimws(sub("^,\\s*", "", from_v))
        
        if (author != "" && !is.null(journal) && journal != "") {
          new_cr <- paste0(author, ", ", year_val, ", ", journal, ", ", from_v_clean)
        } else if (author != "") {
          new_cr <- paste0(author, ", ", year_val, ", ", from_v_clean)
        } else {
          new_cr <- line
        }
        reformatted <- c(reformatted, new_cr)
      } else {
        reformatted <- c(reformatted, line)
      }
    } else {
      reformatted <- c(reformatted, line)
    }
  }
  
  return(paste(reformatted, collapse = "\n"))
}


is_scopus_record <- function(rec) {
  ut <- rec[["UT"]]
  if (is.null(ut)) return(FALSE)
  return(grepl("SCOPUS:", ut))
}


record_to_wos_text <- function(rec) {
  lines <- character(0)
  lines <- c(lines, paste0("PT ", ifelse(is.null(rec[["PT"]]), "J", rec[["PT"]])))
  
  field_order <- c(
    "AU", "AF", "TI", "SO", "LA", "DT", "DE", "ID", "AB",
    "C1", "C3", "RP", "EM", "RI", "OI",
    "FU", "FX", "CR", "NR", "TC", "Z9", "U1", "U2",
    "PU", "PI", "PA", "SN", "EI", "BN",
    "J9", "JI", "PD", "PY", "VL", "IS", "SI",
    "BP", "EP", "AR", "DI", "EA", "PG",
    "PM", "WC", "WE", "SC", "GA", "UT", "OA", "HC", "HP", "DA"
  )
  
  written <- c("PT")
  
  for (tag in field_order) {
    if (!is.null(rec[[tag]])) {
      value <- rec[[tag]]
      
      # Reformat CR for Scopus-origin records
      if (tag == "CR" && is_scopus_record(rec)) {
        value <- reformat_scopus_cr(value)
      }
      
      value_lines <- strsplit(value, "\n")[[1]]
      lines <- c(lines, paste0(tag, " ", value_lines[1]))
      if (length(value_lines) > 1) {
        for (vl in value_lines[2:length(value_lines)]) {
          lines <- c(lines, paste0("   ", vl))
        }
      }
      written <- c(written, tag)
    }
  }
  
  # Write any remaining tags not in field_order
  for (tag in names(rec)) {
    if (!(tag %in% written)) {
      value <- rec[[tag]]
      value_lines <- strsplit(value, "\n")[[1]]
      lines <- c(lines, paste0(tag, " ", value_lines[1]))
      if (length(value_lines) > 1) {
        for (vl in value_lines[2:length(value_lines)]) {
          lines <- c(lines, paste0("   ", vl))
        }
      }
    }
  }
  
  lines <- c(lines, "ER")
  return(paste(lines, collapse = "\n"))
}


merge_records <- function(wos_records, scopus_records, threshold = 0.85, priority = "wos") {
  # Build DOI index for WoS records
  wos_by_doi <- list()
  for (rec in wos_records) {
    doi <- normalize_doi(rec[["DI"]])
    if (!is.na(doi) && doi != "") {
      wos_by_doi[[doi]] <- rec
    }
  }
  
  # Build title list for WoS records
  wos_titles <- lapply(wos_records, function(rec) {
    list(title = normalize_title(rec[["TI"]]), rec = rec)
  })
  
  merged <- wos_records
  duplicates <- list()
  added_from_scopus <- 0
  skipped_doi <- 0
  skipped_title <- 0
  
  for (s_rec in scopus_records) {
    s_doi <- normalize_doi(s_rec[["DI"]])
    is_dup <- FALSE
    
    # Check DOI match
    if (!is.na(s_doi) && s_doi != "" && !is.null(wos_by_doi[[s_doi]])) {
      w_rec <- wos_by_doi[[s_doi]]
      duplicates <- c(duplicates, list(list(
        wos = w_rec, scopus = s_rec, match_type = "DOI", score = 1.0
      )))
      skipped_doi <- skipped_doi + 1
      is_dup <- TRUE
    }
    
    # Check title similarity
    if (!is_dup) {
      s_title <- normalize_title(s_rec[["TI"]])
      if (s_title != "") {
        best_score <- 0
        best_match <- NULL
        
        for (wt in wos_titles) {
          w_title <- wt$title
          if (w_title == "") next
          len_ratio <- nchar(s_title) / max(nchar(w_title), 1)
          if (len_ratio < 0.7 || len_ratio > 1.4) next
          score <- string_similarity(s_title, w_title)
          if (score > best_score) {
            best_score <- score
            best_match <- wt$rec
          }
        }
        
        if (best_score >= threshold) {
          duplicates <- c(duplicates, list(list(
            wos = best_match, scopus = s_rec, match_type = "TITLE", score = best_score
          )))
          skipped_title <- skipped_title + 1
          is_dup <- TRUE
        }
      }
    }
    
    if (!is_dup) {
      merged <- c(merged, list(s_rec))
      added_from_scopus <- added_from_scopus + 1
    }
  }
  
  stats <- list(
    wos_total         = length(wos_records),
    scopus_total      = length(scopus_records),
    duplicates_doi    = skipped_doi,
    duplicates_title  = skipped_title,
    added_from_scopus = added_from_scopus,
    merged_total      = length(merged),
    priority          = priority
  )
  
  return(list(merged = merged, duplicates = duplicates, stats = stats))
}


generate_report <- function(stats, duplicates, report_path) {
  lines <- character(0)
  
  lines <- c(lines, paste(rep("=", 70), collapse = ""))
  lines <- c(lines, "  MERGE & DEDUPLICATION REPORT")
  lines <- c(lines, paste(rep("=", 70), collapse = ""))
  lines <- c(lines, "")
  lines <- c(lines, sprintf("  WoS records:                    %d", stats$wos_total))
  lines <- c(lines, sprintf("  Scopus records:                 %d", stats$scopus_total))
  lines <- c(lines, sprintf("  Total before dedup:             %d", stats$wos_total + stats$scopus_total))
  lines <- c(lines, "")
  lines <- c(lines, sprintf("  Duplicates by DOI:              %d", stats$duplicates_doi))
  lines <- c(lines, sprintf("  Duplicates by title similarity: %d", stats$duplicates_title))
  lines <- c(lines, sprintf("  Total duplicates removed:       %d", stats$duplicates_doi + stats$duplicates_title))
  lines <- c(lines, "")
  lines <- c(lines, sprintf("  Unique from Scopus added:       %d", stats$added_from_scopus))
  lines <- c(lines, sprintf("  Priority on duplicate:          %s", toupper(stats$priority)))
  lines <- c(lines, "")
  lines <- c(lines, sprintf("  FINAL MERGED RECORDS:           %d", stats$merged_total))
  lines <- c(lines, paste(rep("=", 70), collapse = ""))
  lines <- c(lines, "")
  
  if (length(duplicates) > 0) {
    lines <- c(lines, paste(rep("-", 70), collapse = ""))
    lines <- c(lines, "  DUPLICATE DETAILS")
    lines <- c(lines, paste(rep("-", 70), collapse = ""))
    lines <- c(lines, "")
    
    for (i in seq_along(duplicates)) {
      dup <- duplicates[[i]]
      w_title <- substr(gsub("\\s+", " ", ifelse(is.null(dup$wos[["TI"]]), "N/A", dup$wos[["TI"]])), 1, 100)
      s_title <- substr(gsub("\\s+", " ", ifelse(is.null(dup$scopus[["TI"]]), "N/A", dup$scopus[["TI"]])), 1, 100)
      doi <- ifelse(!is.null(dup$wos[["DI"]]), dup$wos[["DI"]],
                    ifelse(!is.null(dup$scopus[["DI"]]), dup$scopus[["DI"]], "N/A"))
      
      match_info <- sprintf("  [%3d] Match: %s", i, dup$match_type)
      if (dup$match_type == "TITLE") {
        match_info <- paste0(match_info, sprintf(" (similarity: %.2f%%)", dup$score * 100))
      }
      lines <- c(lines, match_info)
      lines <- c(lines, sprintf("        DOI:   %s", doi))
      lines <- c(lines, sprintf("        WoS:   %s", w_title))
      lines <- c(lines, sprintf("        Scopus: %s", s_title))
      lines <- c(lines, "")
    }
  }
  
  # Write with BOM
  con <- file(report_path, open = "wb")
  writeBin(charToRaw("\xEF\xBB\xBF"), con)
  writeBin(charToRaw(paste(lines, collapse = "\n")), con)
  close(con)
  
  return(report_path)
}


# ============================================================
# MAIN
# ============================================================
main <- function() {
  dir.create(OUTPUT_FOLDER, showWarnings = FALSE, recursive = TRUE)
  
  cat(sprintf("Loading WoS file: %s\n", WOS_FILE))
  wos_records <- parse_wos_records(WOS_FILE)
  cat(sprintf("  -> %d records\n\n", length(wos_records)))
  
  cat(sprintf("Loading Scopus file: %s\n", SCOPUS_FILE))
  scopus_records <- parse_wos_records(SCOPUS_FILE)
  cat(sprintf("  -> %d records\n\n", length(scopus_records)))
  
  cat("Merging and deduplicating...\n")
  result <- merge_records(
    wos_records, scopus_records,
    threshold = TITLE_SIMILARITY_THRESHOLD,
    priority  = PRIORITY
  )
  
  merged     <- result$merged
  duplicates <- result$duplicates
  stats      <- result$stats
  
  # Write merged file with exact WoS format
  cat(sprintf("\nWriting merged file: %s\n", OUTPUT_FILE))
  con <- file(OUTPUT_FILE, open = "wb")
  writeBin(charToRaw("\xEF\xBB\xBF"), con)  # BOM
  
  write_line <- function(text) {
    writeBin(charToRaw(paste0(text, "\n")), con)
  }
  
  write_line("FN Clarivate Analytics Web of Science")
  write_line("VR 1.0")
  
  for (rec in merged) {
    wos_text <- record_to_wos_text(rec)
    writeBin(charToRaw(paste0(wos_text, "\n\n")), con)
  }
  
  write_line("EF")
  close(con)
  
  report_path <- generate_report(stats, duplicates, REPORT_FILE)
  
  # Validation
  content <- paste(readLines(OUTPUT_FILE, encoding = "UTF-8", warn = FALSE), collapse = "\n")
  pt_count <- length(gregexpr("\nPT ", content)[[1]])
  er_count <- length(gregexpr("\nER\n", content)[[1]])
  
  cat("\n")
  cat(paste(rep("=", 50), collapse = ""), "\n")
  cat("  MERGE SUMMARY\n")
  cat(paste(rep("=", 50), collapse = ""), "\n")
  cat(sprintf("  WoS records:              %d\n", stats$wos_total))
  cat(sprintf("  Scopus records:           %d\n", stats$scopus_total))
  cat(sprintf("  Duplicates (DOI):         %d\n", stats$duplicates_doi))
  cat(sprintf("  Duplicates (title):       %d\n", stats$duplicates_title))
  cat(sprintf("  Unique added from Scopus: %d\n", stats$added_from_scopus))
  cat(sprintf("  TOTAL MERGED:             %d\n", stats$merged_total))
  cat(sprintf("  Validation: PT=%d, ER=%d\n", pt_count, er_count))
  cat(paste(rep("=", 50), collapse = ""), "\n")
  cat(sprintf("\n  Output:  %s\n", OUTPUT_FILE))
  cat(sprintf("  Report:  %s\n", report_path))
}

main()
